### ISSUE TITLE
Editor<->build system interaction

### ISSUE THREAD

**Issue by jeremiedimino** [ocaml/dune#1204]
Currently, in order to be able to quickly jump to build errors in the editor, it is necessary to launch dune from the editor. This is quite annoying, especially since it is often much more convenient to run dune from  a proper shell.

However, it is not clear why we can't have both: i.e. run `dune build` in a normal terminal and have the editor immediately display the build errors and allow the user to cycle through them easily.

**Comment by yminsky** [comment]
Is LSP the right way forward here? I don't know if LSP supports build-system style error reporting, though. I've only seen it highlight errors in the current working buffer, not a workflow where you get a list of extant errors that you can cycle through, potentially across many files.

**Comment by diml** [comment]
Possibly, although I don't know much about LSP. The most difficult part seems to be for the build system to be able to send information to the editor, but without special configuration step from the user.

One simple way would be for the build system to dump the information in special file in the `_build/` directory, and then the editor would just watch this file and report errors. However, this involves watching files the editor, which might not be easy. On Linux, we could also send a signal via D-Bus. Emacs supports watching D-Bus events for instance.

**Comment by diml** [comment]
Looking at this again, we could start with something much simpler: provide a command in dune that returns the list of errors for the current workspace. Then in the editor, we can simply add a function that calls dune to get this list and maintains a list of visited errors, so that we can cycle through them.

With this method, the error buffer won't popup automatically in the editor and it will only appear when the user press a key combination. However, we have a simple way to cycle through errors even if the build wasn't started from within the editor, which is nice.

**Comment by emillon** [comment]
This would also make it possible to highlight errors locations in the terminal (as for dune files).

**Comment by diml** [comment]
@emillon to highlight errors in the terminal, what we need is for dune to understand the locations produced by the compiler, right?

**Comment by diml** [comment]
>  I've tried to do that, but the problem was that the compiler's stdout is directly connected to the terminal, so we'd have to intercept it first.

We already intercept it. We intercept the output of all commands we run as otherwise the output would be mixed up due to running multiple commands in parallel.

@andreypopp I suppose we could use refmterr unconditionally for reason files when it's in the path. Does it work for OCaml as well?

**Comment by andreypopp** [comment]
> I suppose we could use refmterr unconditionally for reason files when it's in the path. 

Sounds good. cc @jordwalke

> Does it work for OCaml as well?

Yes.

**Comment by Khady** [comment]
On Mon, Sep 10, 2018, at 9:18 PM, Etienne Millon wrote:
> Yes, parsing the File: "%s", line %d, characters %d-%d: line should be
> enough. I've tried to do that, but the problem was that the compiler's
> stdout is directly connected to the terminal, so we'd have to
> intercept it first.
It is a bit more complex than that.

See the regexp used by emacs:
https://github.com/emacs-mirror/emacs/blob/master/lisp/progmodes/compile.el#L163-L166
```elisp
    (python-tracebacks-and-caml
     "^[ \t]*File \\(\"?\\)\\([^,\" \n\t<>]+\\)\\1, lines? \\([0-9]+\\)-?\\([0-
     9]+\\)?\\(?:$\\|,\\\(?: characters? \\([0-9]+\\)-?\\([0-9]+\\)?:\\)?\\([ \n]Warning\\(?:
[0-9]+\\)?:\\)?\\)"     2 (3 . 4) (5 . 6) (7))
```

**Comment by Khady** [comment]
On Mon, Sep 10, 2018, at 9:58 PM, Jérémie Dimino wrote:
> @andreypopp[1] I suppose we could use refmterr unconditionally for
> reason files when it's in the path. Does it work for OCaml as well?
Please don't make this unconditional. It means that running dune from
emacs and using integrated system to detect errors/warnings would be
pretty troublesome.

Links:

  1. https://github.com/andreypopp

**Comment by emillon** [comment]
`refmterr` looks great, this is the kind of integration we're after. If it works with ocaml code that's even better.

As for the emacs integration, the point of that issue is to have something that works well with editors. So, either we'll output something that's compatible with what emacs expects (e.g. display the raw one for the editor, then the pretty one), or detect whether we're outputting to a console.

**Comment by emillon** [comment]
FTR, highlighting error locations has been implemented in the compiler itself: https://github.com/ocaml/ocaml/pull/2096

**Comment by joprice** [comment]
@emillon that will be in 4.08? I assume it will be some time before the next release. Could dune provide a temporary solution that allows optionally formatting the output via refmterr?

**Comment by emillon** [comment]
Yes, it has been merged in the 4.08 branch. Many messages have been improved, so I think most of this work will continue in the compiler.

**Comment by rgrinberg** [comment]
I think we can consider as this fixed in 3.0
